File: testbed_php/test/patterns/ClassLevelImmutabilityTest.php
<?php

class ClassLevelImmutabilityTest extends UnitTestCase
{

  public function test_noDeleteForAssociationsAndAttributesOfImmutableClass()
  {
    $associated = new WidgetImmutableB("name");

    $widget = new WidgetImmutableA("Big Widget", $associated);
    $this->assertEqual($widget->getName(), "Big Widget");
    $this->assertEqual($widget->getWidgetImmutableB(), $associated);

    $widget->delete();
    $this->assertEqual($widget->getName(), "Big Widget");
    $this->assertEqual($widget->getWidgetImmutableB(), $associated);
  }

  
  public function test_noDeleteForAssociationsAndAttributesOfSubclassOfImmutableClass()
  {
    $associated = new WidgetImmutableB("name");

    $widget = new WidgetSubclass("Little Widget", "myType", array($associated));
    $this->assertEqual($widget->getName(), "Little Widget");
    $this->assertEqual($widget->getType(), "myType");
    $this->assertEqual($widget->getWidgetImmutableB(0), $associated);

    $widget->delete();
    $this->assertEqual($widget->getName(), "Little Widget");
    $this->assertEqual($widget->getType(), "myType");
    $this->assertEqual($widget->getWidgetImmutableB(0), $associated);
  }


  public function test_mutableClassHasSettersRemoversAndAddMethods()
  {
    $widget = new WidgetMutableB();
    $this->assertTrue($this->objectClassHasSettersAddersOrRemovers($widget));
  }

  
  public function test_unidirectionalManyAssociation()
  {
    $other1 = new ClassOtherclass("otherClass1");
    $other2 = new ClassOtherclass("otherClass2");

    $many = new ClassMany(array());
    $this->assertFalse($many->hasClassOtherclasses());
    $many = new ClassMany(array($other1));
    $this->assertTrue($many->hasClassOtherclasses());
    $many = new ClassMany(array($other1, $other2));
    $this->assertTrue($many->hasClassOtherclasses());
    $this->assertTrue($many->indexOfClassOtherclass($other1) >= 0);
    $this->assertTrue($many->indexOfClassOtherclass($other2) >= 0);
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($many));
  }

  
  public function test_unidirectionalMToNAssociation()
  {
    $other1 = new ClassOtherclass("other1");
    $other2 = new ClassOtherclass("other2");
    $other3 = new ClassOtherclass("other3");
    $other4 = new ClassOtherclass("other4");

    try {
      $mToN = new ClassMToN(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $mToN = new ClassMToN(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $mToN = new ClassMToN(array($other1,$other2));
    $this->assertTrue($mToN->hasClassOtherclasses());
    $this->assertTrue($mToN->indexOfClassOtherclass($other1) >= 0);
    $this->assertTrue($mToN->indexOfClassOtherclass($other2) >= 0);

    $mToN = new ClassMToN(array($other1,$other2,$other3));
    $this->assertTrue($mToN->hasClassOtherclasses());

    try {
      $mToN = new ClassMToN(array($other1,$other2,$other3,$other4));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($mToN));
  }

  
  public function test_unidirectionalMStarAssociation()
  {
    $other1 = new ClassOtherclass("other1");
    $other2 = new ClassOtherclass("other2");
    $other3 = new ClassOtherclass("other3");

    try {
      $mStar = new ClassMStar(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $mStar = new ClassMStar(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $mStar = new ClassMStar(array($other1,$other2));
    $this->assertTrue($mStar->hasClassOtherclasses());
    $this->assertTrue($mStar->indexOfClassOtherclass($other1) >= 0);
    $this->assertTrue($mStar->indexOfClassOtherclass($other2) >= 0);

    $mStar = new ClassMStar(array($other1,$other2,$other3));
    $this->assertTrue($mStar->hasClassOtherclasses());
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($mStar));
  }

  
  public function test_unidirectionalNAssociation()
  {
    $other1 = new ClassOtherclass("other1");
    $other2 = new ClassOtherclass("other2");
    $other3 = new ClassOtherclass("other3");

    try {
      $n = new ClassN(array());
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    try {
      $n = new ClassN(array($other1));
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}

    $n = new ClassN(array($other1,$other2));
    $this->assertTrue($n->hasClassOtherclasses());
    $this->assertTrue($n->indexOfClassOtherclass($other1) >= 0);
    $this->assertTrue($n->indexOfClassOtherclass($other2) >= 0);

    try {
      $n = new ClassN(array($other1,$other2,$other3));
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($n));
  }

  
  public function test_unidirectionalOneAssociation()
  {
    $other = new ClassOtherclass("otherClass");

    $one = new ClassOne($other);
    $this->assertTrue($one->getClassOtherclass() == $other);

    try {
      $one = new ClassOne(null);
      $this->fail("Constructor should have thrown Exception");
    } catch (Exception $ex) {}
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($one));  
  }

  
  public function test_unidirectionalOptionalNAssociation()
  {
    $other1 = new ClassOtherclass("other1");
    $other2 = new ClassOtherclass("other2");
    $other3 = new ClassOtherclass("other3");

    $optN = new ClassOptionalN(array());
    $this->assertFalse($optN->hasClassOtherclasses());
    
    $optN = new ClassOptionalN(array($other1));
    $this->assertTrue($optN->hasClassOtherclasses());
    $this->assertTrue($optN->indexOfClassOtherclass($other1) >= 0);

    $optN = new ClassOptionalN(array($other1,$other2));
    $this->assertTrue($optN->hasClassOtherclasses());
    $this->assertTrue($optN->indexOfClassOtherclass($other1) >= 0);
    $this->assertTrue($optN->indexOfClassOtherclass($other2) >= 0);

    try {
      $optN = new ClassOptionalN(array($other1,$other2,$other3));
      $this->fail("Constructor should have thrown RuntimeException");
    } catch (Exception $ex) {}

    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($optN));   
  }

  
  public function test_unidirectionalOptionalOneAssociation()
  {
    $other = new ClassOtherclass("otherClass");

    $optOne = new ClassOptionalOne($other);
    $this->assertTrue($optOne->getClassOtherclass() == $other);

    $optOne = new ClassOptionalOne(null);
    $this->assertTrue($optOne->getClassOtherclass() == null);
    
    $this->assertFalse($this->objectClassHasSettersAddersOrRemovers($optOne));    
  }
  

  private function objectClassHasSettersAddersOrRemovers($obj)
  {
    $hasMethods = false;

    $clazz = new ReflectionClass(get_class($obj));
    $methods = $clazz->getMethods();

    foreach ($methods as $m)
    {
      if ((preg_match('/^(set|add|remove)[A-Z]+[a-zA-Z]*$/', $m->getName()) > 0) && $m->isPublic())
      {
        $hasMethods = true;
        break;
      }
    }
    return $hasMethods;
  }
}

?>

File: testbed_ruby/test/patterns/class_level_immutability_test.rb
require 'test/unit'
require './src-gen-umple/widget_immutable_a'
require './src-gen-umple/widget_immutable_b'
require './src-gen-umple/widget_subclass'
require './src-gen-umple/widget_mutable_b'
require './src-gen-umple/class_many'
require './src-gen-umple/class_m_to_n'
require './src-gen-umple/class_m_star'
require './src-gen-umple/class_n'
require './src-gen-umple/class_one'
require './src-gen-umple/class_optional_n'
require './src-gen-umple/class_optional_one'
require './src-gen-umple/class_otherclass'

module CruisePatternsTest
class ClassLevelImmutabilityTest < Test::Unit::TestCase

  def test_noDeleteForAssociationsAndAttributesOfImmutableClass
    associated = WidgetImmutableB.new("name")

    widget = WidgetImmutableA.new("Big Widget",associated)
    assert_equal(widget.get_name, "Big Widget")
    assert_equal(widget.get_widgetImmutableB, associated)

    widget.delete
    assert_equal(widget.get_name, "Big Widget")
    assert_equal(widget.get_widgetImmutableB, associated)
  end
  

  def test_noDeleteForAssociationsAndAttributesOfSubclassOfImmutableClass
    associated = WidgetImmutableB.new("name")

    widget = WidgetSubclass.new("Little Widget","myType",[associated])
    assert_equal(widget.get_name, "Little Widget")
    assert_equal(widget.get_type, "myType")
    assert_equal(widget.get_widgetImmutableB(0), associated)

    widget.delete
    assert_equal(widget.get_name, "Little Widget")
    assert_equal(widget.get_type, "myType")
    assert_equal(widget.get_widgetImmutableB(0), associated)
  end


  def test_mutableClassHasSettersRemoversAndAddMethods
    widget = WidgetMutableB.new
    assert_equal(true,object_class_has_adders_or_removers(widget))
  end
  

  def test_unidirectionalManyAssociation
    other1 = ClassOtherclass.new("otherClass")
    other2 = ClassOtherclass.new("otherClass")

    many = ClassMany.new([])
    assert_equal(false,many.has_classOtherclasses)
    many = ClassMany.new([other1])
    assert_equal(true,many.has_classOtherclasses)
    many = ClassMany.new([other1, other2])
    assert_equal(true,many.has_classOtherclasses)
    assert_equal(true,many.index_of_classOtherclass(other1) >= 0)
    assert_equal(true,many.index_of_classOtherclass(other2) >= 0)
    assert_equal(false,object_class_has_adders_or_removers(many))

    assert_equal(false,many.set_classOtherclasses([other1,other2]))
  end  


  def test_unidirectionalMToNAssociation
    other1 = ClassOtherclass.new("other")
    other2 = ClassOtherclass.new("other")
    other3 = ClassOtherclass.new("other")
    other4 = ClassOtherclass.new("other")

    assert_raise RuntimeError do 
      mToN = ClassMToN.new([])
    end
    
    assert_raise RuntimeError do
      mToN = ClassMToN.new([other1])
    end    

    mToN = ClassMToN.new([other1,other2])
    assert_equal(true,mToN.has_classOtherclasses)
    assert_equal(true,mToN.index_of_classOtherclass(other1) >= 0)
    assert_equal(true,mToN.index_of_classOtherclass(other2) >= 0)

    assert_equal(false,mToN.set_classOtherclasses([other1,other2]))

    mToN = ClassMToN.new([other1,other2,other3])
    assert_equal(true,mToN.has_classOtherclasses)

    assert_raise RuntimeError do
      mToN = ClassMToN.new([other1,other2,other3,other4])
    end    

    assert_equal(false,object_class_has_adders_or_removers(mToN))
  end  


  def test_unidirectionalMStarAssociation
    other1 = ClassOtherclass.new("other")
    other2 = ClassOtherclass.new("other")
    other3 = ClassOtherclass.new("other")

    assert_raise RuntimeError do 
      mStar = ClassMStar.new([])
    end

    assert_raise RuntimeError do 
      mStar = ClassMStar.new([other1])
    end

    mStar = ClassMStar.new([other1,other2])
    assert_equal(true,mStar.has_classOtherclasses)
    assert_equal(true,mStar.index_of_classOtherclass(other1) >= 0)
    assert_equal(true,mStar.index_of_classOtherclass(other2) >= 0)

    assert_equal(false,mStar.set_classOtherclasses([other1,other2]))

    mStar = ClassMStar.new([other1,other2,other3])
    assert_equal(true,mStar.has_classOtherclasses)
    
    assert_equal(false,object_class_has_adders_or_removers(mStar))
  end  


  def test_unidirectionalNAssociation
    other1 = ClassOtherclass.new("other")
    other2 = ClassOtherclass.new("other")
    other3 = ClassOtherclass.new("other")

    assert_raise RuntimeError do 
      n = ClassN.new([])
    end

    assert_raise RuntimeError do 
      n = ClassN.new([other1])
    end

    n = ClassN.new([other1,other2])
    assert_equal(true,n.has_classOtherclasses)
    assert_equal(true,n.index_of_classOtherclass(other1) >= 0)
    assert_equal(true,n.index_of_classOtherclass(other2) >= 0)

    assert_equal(false,n.set_classOtherclasses([other1,other2]))

    assert_raise RuntimeError do 
      n = ClassN.new([other1,other2,other3])
    end
    
    assert_equal(false,object_class_has_adders_or_removers(n))
  end


  def test_unidirectionalOneAssociation
    other = ClassOtherclass.new("otherClass")

    one = ClassOne.new(other)
    assert_equal(true,one.get_classOtherclass == other)

    assert_raise RuntimeError do 
      one = ClassOne.new(nil)
    end
    
    assert_equal(false,object_class_has_adders_or_removers(one))
  end  


  def test_unidirectionalOptionalNAssociation
    other1 = ClassOtherclass.new("other")
    other2 = ClassOtherclass.new("other")
    other3 = ClassOtherclass.new("other")

    optN = ClassOptionalN.new([])
    assert_equal(false,optN.has_classOtherclasses)
    
    optN = ClassOptionalN.new([other1])
    assert_equal(true,optN.has_classOtherclasses)
    assert_equal(true,optN.index_of_classOtherclass(other1) >= 0)

    optN = ClassOptionalN.new([other1,other2])
    assert_equal(true,optN.has_classOtherclasses)
    assert_equal(true,optN.index_of_classOtherclass(other1) >= 0)
    assert_equal(true,optN.index_of_classOtherclass(other2) >= 0)

    assert_equal(false,optN.set_classOtherclasses([other1,other2]))

    assert_raise RuntimeError do 
      optN = ClassOptionalN.new([other1,other2,other3])
    end

    assert_equal(false,object_class_has_adders_or_removers(optN))
  end


  def test_unidirectionalOptionalOneAssociation
    other = ClassOtherclass.new("otherClass")

    optOne = ClassOptionalOne.new(other)
    assert_equal(true,optOne.get_classOtherclass == other)

    optOne = ClassOptionalOne.new(nil)
    assert_equal(true,optOne.get_classOtherclass == nil)
    
    assert_equal(false,object_class_has_adders_or_removers(optOne))
  end


  def object_class_has_adders_or_removers(obj)
    hasMethods = false
    methods = obj.class.public_instance_methods
    setterNameRegex = /\A(add|remove)_[a-z]+[a-zA-Z]*\Z/

    methods.each do |m|
      if ((m =~ setterNameRegex) == 0)
        hasMethods = true
        break
      end
    end
    hasMethods
  end

end
end
